src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage_layout.h

```diff
@@ -0,0 +1,437 @@
+  template <KeyT KT>
+  static void append_key(
+      NodeExtentMutable& mut, const full_key_t<KT>& key, char*& p_append) {
+    append_key(mut, key_t::template from_key<KT>(key), p_append);
+  }
+  static void append_offset(
+      NodeExtentMutable& mut, node_offset_t offset_to_right, char*& p_append);
+
+  node_header_t header;
+  num_keys_t num_keys = 0u;
+  SlotType slots[];
+} __attribute__((packed));
+using node_fields_0_t = _node_fields_013_t<slot_0_t>;
+using node_fields_1_t = _node_fields_013_t<slot_1_t>;
+
+// internal/leaf node N2
+struct node_fields_2_t {
+  // TODO: decide by NODE_BLOCK_SIZE, sizeof(node_off_t), sizeof(laddr_t)
+  // and the minimal size of variable_key.
+  using num_keys_t = uint8_t;
+  using key_t = ns_oid_view_t;
+  using key_get_type = key_t;
+  static constexpr field_type_t FIELD_TYPE = field_type_t::N2;
+  static constexpr node_offset_t SIZE = NODE_BLOCK_SIZE;
+  static constexpr node_offset_t HEADER_SIZE =
+    sizeof(node_header_t) + sizeof(num_keys_t);
+  static constexpr node_offset_t ITEM_OVERHEAD = sizeof(node_offset_t);
+
+  bool is_level_tail() const { return header.get_is_level_tail(); }
+  node_offset_t total_size() const { return SIZE; }
+  key_get_type get_key(size_t index) const {
+    assert(index < num_keys);
+    node_offset_t item_end_offset =
+      (index == 0 ? SIZE : offsets[index - 1]);
+    assert(item_end_offset <= SIZE);
+    const char* p_start = fields_start(*this);
+    return key_t(p_start + item_end_offset);
+  }
+  node_offset_t get_key_start_offset(size_t index) const {
+    assert(index <= num_keys);
+    auto offset = HEADER_SIZE + sizeof(node_offset_t) * num_keys;
+    assert(offset <= SIZE);
+    return offset;
+  }
+  node_offset_t get_item_start_offset(size_t index) const {
+    assert(index < num_keys);
+    auto offset = offsets[index];
+    assert(offset <= SIZE);
+    return offset;
+  }
+  const void* p_offset(size_t index) const {
+    assert(index < num_keys);
+    return &offsets[index];
+  }
+  node_offset_t get_item_end_offset(size_t index) const {
+    return index == 0 ? SIZE : get_item_start_offset(index - 1);
+  }
+  template <node_type_t NODE_TYPE>
+  node_offset_t free_size_before(size_t index) const {
+    auto range = fields_free_range_before<NODE_TYPE>(*this, index);
+    return range.end - range.start;
+  }
+
+#if 0
+  template <node_type_t NODE_TYPE>
+  void fill_unused(NodeExtentMutable& mut) const {
+    auto range = fields_free_range_before<NODE_TYPE>(*this, num_keys);
+    for (auto i = range.start; i < range.end; ++i) {
+      mut.copy_in_relative(i, uint8_t(0xc5));
+    }
+  }
+
+  template <node_type_t NODE_TYPE>
+  void validate_unused() const {
+    auto range = fields_free_range_before<NODE_TYPE>(*this, num_keys);
+    for (auto i = fields_start(*this) + range.start;
+         i < fields_start(*this) + range.end;
+         ++i) {
+      assert(*i == char(0xc5));
+    }
+  }
+#endif
```

Comment:
Probably remove?

Reply:
removed in https://github.com/ceph/ceph/pull/36986/commits/7ed1df465cf8059fab7ba0910e4c75b72a68ce7a